src/serve.rs

```diff
@@ -55,186 +52,218 @@ pub enum Error {
+        TransactionEnvelope::TxV0(envelope) => envelope.tx.operations,
+        TransactionEnvelope::Tx(envelope) => envelope.tx.operations,
+        TransactionEnvelope::TxFeeBump(envelope) => {
+            let FeeBumpTransactionInnerTx::Tx(tx_envelope) = envelope.tx.inner_tx;
+            tx_envelope.tx.operations
+        }
+    };
+    if ops.len() != 1 {
+        return Err(Error::Xdr(XdrError::Invalid));
+    }
+    let op = ops.first().ok_or(Error::Xdr(XdrError::Invalid))?;
+    let body = if let OperationBody::InvokeHostFunction(b) = &op.body {
+        b
+    } else {
+        return Err(Error::Xdr(XdrError::Invalid));
+    };
+
+    if body.function != HostFunction::Call {
+        return Err(Error::Xdr(XdrError::Invalid));
+    };
 
-    // Initialize storage and host
-    // TODO: allow option to separate input and output file
-    let ledger_entries = snapshot::read(ledger_file)?;
+    if body.parameters.len() < 2 {
+        return Err(Error::Xdr(XdrError::Invalid));
+    };
 
-    let snap = Rc::new(snapshot::Snap {
-        ledger_entries: ledger_entries.clone(),
-    });
-    let mut storage = Storage::with_recording_footprint(snap);
-    let contents = utils::get_contract_wasm_from_storage(&mut storage, contract_id)?;
-    let h = Host::with_storage(storage);
-
-    let vm = Vm::new(&h, [0; 32].into(), &contents).unwrap();
-    let inputs = match contractspec::function_spec(&vm, func) {
-        Some(s) => s.inputs,
-        None => {
-            return Err(Error::FunctionNotFoundInContractSpec);
-        }
+    let contract_xdr = body
+        .parameters
+        .get(0)
+        .ok_or(Error::Xdr(XdrError::Invalid))?;
+    let method_xdr = body
+        .parameters
+        .get(1)
+        .ok_or(Error::Xdr(XdrError::Invalid))?;
+    let (_, params) = body.parameters.split_at(2);
+
+    let contract_id: [u8; 32] = if let ScVal::Object(Some(ScObject::Bytes(bytes))) = contract_xdr {
+        bytes
+            .as_slice()
+            .try_into()
+            .map_err(|_| Error::Xdr(XdrError::Invalid))?
+    } else {
+        return Err(Error::Xdr(XdrError::Invalid));
     };
 
-    // re-assemble the args, to match the order given on the command line
-    let args: Vec<ScVal> = if args_xdr.is_empty() {
-        args.iter()
-            .zip(inputs.iter())
-            .map(|(a, input)| strval::from_json(a, &input.type_))
-            .collect::<Result<Vec<_>, _>>()?
+    // TODO: Figure out and enforce the expected type here. For now, handle both a symbol and a
+    // binary. The cap says binary, but other implementations use symbol.
+    let method: String = if let ScVal::Object(Some(ScObject::Bytes(bytes))) = method_xdr {
+        bytes
+            .try_into()
+            .map_err(|_| Error::Xdr(XdrError::Invalid))?
+    } else if let ScVal::Symbol(bytes) = method_xdr {
+        bytes
+            .try_into()
+            .map_err(|_| Error::Xdr(XdrError::Invalid))?
     } else {
```

Comment:
I thought we had added conversions for these that were easier, like scval.try_into()? for string, but I might be wrong.

Reply:
Oh maybe it wasn't in the older version of the sdk, but they're there now?

Reply:
Hmmm... Edit still says there's no ScVal -> String TryInto implementation.